core/src/socket.rs

```diff
@@ -0,0 +1,243 @@
+                    actions.push(SocketAction::Connect(handle, success));
+                    if !success {
+                        continue;
+                    }
+                }
+                None => continue,
+
+                _ => {}
+            }
+
+            if let Some(received) = internal.poll() {
+                actions.push(SocketAction::Data(handle, received));
+            } else if matches!(is_connected, Some(false) | None) {
+                actions.push(SocketAction::Close(handle));
+                continue;
+            }
+
+            if let Some(to_send) = send_buffer.pop_front() {
+                internal.send(to_send);
+            }
+        }
+
+        if actions.is_empty() {
+            return;
+        }
+
+        let mut activation =
+            Activation::from_stub(uc.reborrow(), ActivationIdentifier::root("[XMLSocket]"));
+
+        for action in actions {
+            match action {
+                SocketAction::Connect(handle, success) => {
+                    let arena = &mut activation.context.xml_sockets.0;
+                    let target = if success {
+                        arena
+                            .get(handle)
+                            .expect("only valid handles in SocketAction")
+                            .target
+                    } else {
+                        arena
+                            .remove(handle)
+                            .expect("only valid handles in SocketAction")
+                            .target
+                    };
+
+                    let _ = TObject::call_method(
+                        &target,
+                        "onConnect".into(),
+                        &[success.into()],
+                        &mut activation,
+                        ExecutionReason::FunctionCall,
+                    );
+                }
+                SocketAction::Close(handle) => {
+                    let target = activation
+                        .context
+                        .xml_sockets
+                        .0
+                        .remove(handle)
+                        .expect("only valid handles in SocketAction")
+                        .target;
+
+                    let _ = TObject::call_method(
+                        &target,
+                        "onClose".into(),
+                        &[],
+                        &mut activation,
+                        ExecutionReason::FunctionCall,
+                    );
+                }
+                SocketAction::Data(handle, data) => {
+                    let target = activation
+                        .context
+                        .xml_sockets
+                        .0
+                        .get(handle)
+                        .expect("only valid handles in SocketAction")
+                        .target;
+
+                    let data =
+                        AvmString::new_utf8(activation.context.gc_context, UTF_8.decode(&data).0);
```

Comment:
the utf8 decode here seems to break some protocols,
seems to work better with `let data = AvmString::new_utf8_bytes(activation.context.gc_context, &data);`